Step3

sirupe opened the pull request.

sirupe committed:
feature(readme.md)
(1) Added requirement at readme.md

sirupe committed:
feature(Accumulator/AdditionSplitor/Fomula)
(1) Created Accumulator class with test code.
(2) Created AdditionSplitor class with test code.
(3) Created Fomula class with test code.

sirupe committed:
feature(Accumulator/AdditionSplitor/Fomula)
(1) Created Accumulator class with test code.
(2) Created AdditionSplitor class with test code.
(3) Created Fomula class with test code.

Commit:
feature(Formula)
(1) Added getCustomSeparator method in Formula method
(2) Added requirements to README.md

Commit:
feature(PatternFormula/Separators/Spliter)
(1) Removed Formula, AdditionSplitor class
(2) Created Separators, Spliter class
(3) Added split formula logic to Spliter and Separators
(4) Added test class

Commit:
feature(PatternFormula)
(1) Added logic to separate formula in PatternFormula class

Commit:
feature(PatternFormula)
(1) Added logic to PatternFormula to separate custom separator
(2) Added logic to PatternFormula to separate number formula

Commit:
feature(PatternFormula/Spliter)
(1) Added conditional canSeparateCustomSeparator class in PatternFormula class
(2) Added test

Commit:
feature
(1) Added validate logic to Accumulator class.
(2) Created NumberFormula class

Commit:
refactor

Commit:
refactor

Commit:
refactor
(1) Accumulator class : int 배열을 받아 계산하는 accumulate 메소드를 삭제하고 문자열 배열을 받아 진행하는 메소드에서 숫자인지 검증하는 메소드의 이름을 isSignedNumeric 으로 변경하였습니다.
(2) test코드의 상수를 없애고 각 테스트 코드에 나타나도록 수정하였습니다.

Commit:
refactor
(1) 문자열 계산기 패키지 생성하여 관련 클래스 해당 패키지로 이동

Commit:
feature(LottoNumber)
(1) LottoNumber 클래스 생성 및 테스트 코드 작성

Commit:
feature(LottoPrice)
(1) LottoPrice 클래스 및 테스트 코드 추가
(2) LottoPrice 에 구매 가능 횟수 추가

Commit:
feature(LottoPrice)
(1) 기본 구조 생성 (어플리케이션, 컨트롤러, 뷰)

Commit:
feature(Inputer, Outer, RandomNumberMaker)
(1) 뷰 관련 입력 출력 인터페이스 및 클래스 추가
(2) 랜덤 번호 생성기 추가

Commit:
feature(Lotto, Lottos)
(1) Lotto (로또 한 개), Lottos (로또 여러개) 클래스 추가
(2) Lottos 클래스에 로또 구매 로직 추가 및 테스트

Commit:
feature(LottoNumbersMaker)
(1) RandomNumberMaker -> LottoNumberMake 이름 변경 및 구현

Commit:
feature(LottoNumbersMaker)
(1) 랜덤 로또번호 구해오기 로직 수정
(2) Lotto 클래스 길이체크 메소드, 중복체크 메소드 추가
(3) LottoNumber 클래스에 sort 하기 위해 Comparable 인터페이스 구현
(4) Lottos 뽑힌 번호들 출력 추가

sirupe committed:
feature(Lotto/LottoNumber/Lotts)
(1) Lotto 생성시 validation 체크 추가
(2) 자동생성된 로또번호 출력을 위한 toString override

sirupe committed:
feature(WinLotto/WinNumbers)
(1) 스캐너 에러사항 수정
(2) 당첨번호 입력 및 생성 로직 추기

sirupe committed:
feature(BoughtLottos/CashPayments/Prize/WinNumbersCount)
(1) 구매한 로또번호와 로또 금액을 같이 가지고 있는 클래스 추가 (BoughtLottos)
(2) 구매한 로또 금액정보 클래스 추가 (CashPayments)
(3) 로또 당첨 금액정보 클래스 추가 (Prize)
(4) 로또 당첨 금액별 횟수 저장 클래스 추가 (WinNumbersCount)
(5) BoughtLottos 클래스를 추가함에 따른 기능 이동(Lottos -> BoughtLottos) 및 테스트 코드 이동

sirupe committed:
작업내용 중간저장

Commit:
feature(로또 로직 완료)
(1) 수익률 구하기 로직 추가
(2) 결과 출력 로직 추가.
(3) 추가에 따른 테스트 코드 추가

Commit:
refactor
(1) 사용하지 않는 변수 및 메소드 삭제
(2) 변수에 final 을 사용하도록 변경
(3) 클래스명 및 변수명 변경

Commit:
refactor : 금액부분 언더바(_) 표시

sirupe committed:
refactor
(1) accumulate 정규식 수정
(2) accumulate 테스트 코드 수정
(3) lotto/ConsoleInputer >> scanner static 으로 수정
(4) BuyingLottos >> buyLottos 메소드 생성자에서 실행하도록 변경
(5) Prize >> 당첨 및 금액 0 원 값 추가

sirupe committed:
refactor
readme.md 내용 추가

Commit:
feature(step3 완료)
(1) InputView >> 보너스 볼 받기 추가
(2) BonusNumber >> 보너스 번호 클래스 추가
(3) 결과 클래스들 >> BonusNumber 포함하여 결과 출력할 수 있도록 수정
(4) Prize >> 2등에 해당하는 당첨정보 추가, 결과 멘트부 Prize 로 이동

Commit:
refactor
(1) 삭제가 필요한 테스트 삭제

sirupe committed:
refactor
카멜케이스 적용 안된 부분 카멜케이스로 변경

sirupe committed:
Merge branch 'step3' of github.com:sirupe/java-lotto into step4

imcts11 commented on src/main/java/accumulate/Accumulator.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Arrays;
+
+public class Accumulator {
+    private final static String NUMBER_REGEX = "^[0-9]*";
```
만약 표현하고자 함이 숫자만인거라면 `"^[0-9]+$"` 이렇게 되어야 할 것 같음. 
`*`은 0개이상이고 `+`는 1개이상. `^`는 숫자로 시작하고 `$`숫자로 종료되어야함을 의미함.

imcts11 replied:
현재 정규식은 숫자로 시작되고 숫자로 종료되는 것을 검증하진 않음.

sirupe committed:
중간저장

Commit:
feature(4단계구현)
(1) LottoController -> 수동 건수와 번호를 입력받는 view 를 호출하여 모델에 전달
(2) BuyingLottos -> 수동 건수와 번호를 저장하고 계산하는 로직 추가
(3) SelfInputCount -> 수동 건수 입력을 위해 새로운 클래스 추가
(4) InputView -> 수동 건수와 번호를 입력받는 질문 추가
(5) 추가된 클래스에 대해 메소드 추가
(6) 변경내역에 따른 테스트케이스 수정

sirupe committed:
refactor
(1) 사용하지 않는 클래스 및 메소드 삭제

imcts11 commented on src/main/java/accumulate/Accumulator.java:
```diff
@@ -0,0 +1,22 @@
+            throw new RuntimeException(ErrorMessage.INCORRECT_VALUE.getMessage());
+        }
+        return Arrays.stream(numbers)
+          .mapToInt(number -> Integer.parseInt(number))
```
.mapToInt(Integer::parseInt)로 축약 가능.

imcts11 commented on src/main/java/accumulate/PatternFormula.java:
```diff
@@ -0,0 +1,71 @@
+    }
+    
+    private static boolean canSeparateCustomSeparator(String separator) {
+        boolean correctLength = separator.length() == SEPARETOR_MAX_LENGTH;
```
명시적 final 추가.

imcts11 commented on src/main/java/accumulate/PatternFormula.java:
```diff
@@ -0,0 +1,71 @@
+    private final static String CUSTOM_SEPARATOR_REGEX = "(\\/\\/.)";
+    private final static int SEPARATOR_INDEX = 0;
+    private final static int NUMBER_FORMULA_INDEX = 1;
+    private final static int SEPARETOR_MAX_LENGTH = 3;
```
오타 수정. SEPARETOR_MAX_LENGTH -> SEPARATOR_MAX_LENGTH

imcts11 commented on src/main/java/lotto/common/ConsoleInputer.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Scanner;
+
+public class ConsoleInputer implements Inputer {
```
Ìnputer`가 있어? `Inputor 아니야?

imcts11 commented on src/main/java/lotto/common/ConsoleInputer.java:
```diff
@@ -0,0 +1,23 @@
+    private static Scanner scanner;
+    
+    static {
+        scanner = new Scanner(System.in);
```
`static`영역을 초기화 하는 영역을 사용했는데. 보통 스태틱 초기화 영역을 사용할때에는 어떠한 로직이 필요하거나 특정 조건에 따라서 값이 다르게 할당되어야 할때 사용하게 됨. 근데 여기서는 그냥 스캐너를 생성하고 말아야 하는데, 왜 굳이 static 초기화 블록을 사용해야 하는지 의문임. 불필요한 context를 사용하게 만드는 역할. 제거.

imcts11 commented on src/main/java/lotto/common/ConsoleInputer.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public int inputNumber() {
+        int number = scanner.nextInt();
+        scanner.nextLine();
```
질문: scanner.nextLine()을 왜 두번 호출하고, 그 값을 할당하지 않는지?

imcts11 replied:
만약 이유가 있어서 해당 코드를 작성 했다면 주석으로 왜 이렇게 작성 했는지에 대해 남겨두어야 함.
/**
* @issues
* 내용
*/

imcts11 commented on src/main/java/lotto/view/ResultView.java:
```diff
@@ -0,0 +1,31 @@
+        this.outer = outer;
+    }
+    
+    public void printLottos(final BuyingLottos lottos) {
```
Lottoes가 맞을거 같은데.

imcts11 commented on src/main/java/lotto/LottoApplication.java:
```diff
@@ -0,0 +1,20 @@
+        final ResultView consoleResultView = new ResultView(outer);
+        final LottoController controller = new LottoController(inputView, consoleResultView);
+        
+        controller.runBuyingLottoProcess();
```
정확하게 이 메소드 두가지가 `public`이어야 할 이유가 전혀 없어보임. 
1. runBuyingLottoProcess보다 runResultProcess가 먼저 실행되면 어떻게 됨 ?
2. 해당 두 메소드는 `private`으로 만들어 내부로 숨기고 (로직을 분리한 거니까) 생성자에서 호출하면 됨.
3. 지금처럼 controller가 초기화되고난 이후에 어떤 추가적인 작업이 있다면 public 메소드는 1개만 외부로 유출하면 될 것. 
- 예시) controller.run()  
- run 내부에서 저 두 메소드를 호출. 
- 이렇게 되면 생성 / 실행 단계가 나누어지는게 가능. 외부에서 메소드를 혼재하여 호출할 일이 없어짐.

imcts11 commented on src/main/java/lotto/domain/BuyingLottos.java:
```diff
@@ -0,0 +1,73 @@
+            throw new IllegalArgumentException(ErrorMessage.NOT_ENOUGH_CASH_PAYMENT.message());
+        }
+        this.cashPayments = new CashPayments(cashPayments);
+        buyableCount = new BuyableCount(cashPayments / MIN_BUYABLE_PAYMENT);
```
`BuyableCount`는 왜 생성자에서 유효성 검증을 처리하지 않고, 계산된 값을 넘겨주고, 심지어 생성자 내부에서는 그냥 값을 받아서 저장만 하는 VO로 사용하고 있는지. 
사실 `BuyableCount`내부에서 cashPayments 값을 받아서 / MIN_BUYABLE_PAYMENT로 진행해야 하지 않아? 
그리고 그 상태값을 기준으로 판단해야 한다면 해당 클래스 내부에 isBuyable()이 있어야 하겠지.

imcts11 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,56 @@
+    
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
중괄호를 생략하지 말자.

imcts11 replied:
생략할거면 모든 단항 if문에도 생략하자.

imcts11 commented on src/main/java/lotto/domain/WinNumbersCount.java:
```diff
@@ -0,0 +1,46 @@
+    public int addWinCount(final int matchCount, final boolean isMatchedBonus) {
+        final Prize matchPrize = Prize.getPrize(matchCount, isMatchedBonus);
+        if (matchPrize == null) {
+            return 0;
```
DEFAULT_WIN_COUNT로 변경

imcts11 commented on src/test/java/accumulate/AccumulatorTest.java:
```diff
@@ -0,0 +1,40 @@
+    @DisplayName("분리된 숫자의 합을 반환.")
+    void accumulateTest() {
+        //Given
+        String[] numbers = {"1", "2", "3"};
```
// Given이 끝나고 
코드
코드
(공백)
// When
코드
코드
공백으로 처리

```javascript
// Given
let a = 3;

// When
a += 3

// Then
assert(a).toEquals(6)
```
